src/main/java/com/uipath/uipathpackage/UiPathDeploy.java

```diff
@@ -106,37 +108,46 @@ public BuildStepMonitor getRequiredMonitorService() {
-        listener.getLogger().println("Opening Powershell Session");
+        FilePath tempRemoteDir = tempDir(workspace);
         try {
-            String importModuleCommands = util.importModuleCommands(listener, env);
-            String deployPackCommand = String.format("Deploy -orchestratorAddress %s -tenant %s -username %s -password %s -packagePath %s -authType UserPass", escapePowerShellString(orchestratorAddress), orchestratorTenantFormatted, username, password, packagePathFormatted);
-            util.execute(listener, importModuleCommands, deployPackCommand);
-            listener.getLogger().println("Exiting Powershell Session");
-        } catch (IOException | URISyntaxException e) {
+            tempRemoteDir.mkdirs();
+            EnvVars envVars = run.getEnvironment(listener);
+            FilePath packagePathFormatted = new FilePath(new File(envVars.expand(packagePath.trim())));
+            if (packagePathFormatted.isDirectory()) {
+                packagePathFormatted.copyRecursiveTo(tempRemoteDir);
+            } else {
+                packagePathFormatted.getParent().copyRecursiveTo(packagePathFormatted.getName(), tempRemoteDir);
+            }
+            ResourceBundle rb = ResourceBundle.getBundle("config");
+            String orchestratorTenantFormatted = envVars.expand(orchestratorTenant.trim()).isEmpty() ? util.getValue(rb, "UiPath.DefaultTenant") : util.escapePowerShellString(envVars.expand(orchestratorTenant.trim()));
+            StandardUsernamePasswordCredentials cred = CredentialsProvider.findCredentialById(credentialsId, StandardUsernamePasswordCredentials.class, run, Collections.emptyList());
+            if (cred == null || cred.getUsername().isEmpty() || cred.getPassword().getPlainText().isEmpty()) {
+                throw new AbortException("Invalid credentials");
+            }
+            String username = util.escapePowerShellString(cred.getUsername());
+            String password = util.escapePowerShellString(cred.getPassword().getPlainText());
+            String importModuleCommands = util.importModuleCommands(tempRemoteDir, listener, envVars);
+            String deployPackCommand = String.format("Deploy -orchestratorAddress %s -tenant %s -username %s -password %s -packagePath %s -authType UserPass", util.escapePowerShellString(orchestratorAddress), orchestratorTenantFormatted, username, password, util.escapePowerShellString(tempRemoteDir.getRemote()));
+            if (!(new PowerShell(util.getCommand(importModuleCommands, deployPackCommand)).perform((AbstractBuild<?, ?>) run, launcher, listener))) {
+                throw new AbortException("Failed to execute powershell session while importing the module and deploying. Command : " + importModuleCommands + " " + deployPackCommand);
+            }
+        } catch (URISyntaxException e) {
             e.printStackTrace(listener.getLogger());
             throw new AbortException(e.getMessage());
+        } finally {
```

Comment:
Please put the deleteRecursive call itself in a try catch block to prevent a potential exception from a finally block.

Reply:
deleteRecursive will throw IOException and InterruptedException which is already handled by the calling method and build will fail, this method has been marked with throws IOException and Interrupted Excpetion, so no need to handle it separately. Wrapping it in try catch, and again throwing it will mark it as probable bug in analysis as we shouldn't throw in finally

Reply:
Ideally finally block should never be throwing exceptions intentionally or unintentionally. Why do you need to fail in case deleting the temp folder fails? That is what will happen with your current changes